src/blkdev.c

```diff
@@ -0,0 +1,384 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Simon Kuenzer <simon.kuenzer@neclab.eu>
+ *          Fabrice Buoro <fabrice@tarides.com>
+ *          Samuel Hym <samuel@tarides.com>
+ *
+ * Copyright (c) 2013-2017, NEC Europe Ltd., NEC Corporation.
+ *               2024-2025, Tarides.
+ *               All rights reserved.
+*/
+
+#include "block.h"
+#include "result.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "yield.h"
+
+block_t blocks[MAX_BLK_DEVICES] = {0};
+
+static token_t *acquire_token(block_t *block)
+{
+  token_t *token;
+
+  for (unsigned int i = 0; i < MAX_BLK_TOKENS; i++) {
+    unsigned long pos = 1L << i;
+    if ((block->infly & pos) == 0) {
+      token = &block->tokens[i];
+      block->infly |= pos;
+      return token;
+    }
+  }
+  assert(uk_blkreq_is_done(&token->req));
+  return NULL;
+}
+
+static token_id_t token_id(block_t *block, token_t *token)
+{
+  return token - block->tokens;
+}
+
+static void release_token(block_t *block, token_t *token)
+{
+  token_id_t id = token_id(block, token);
+
+  unsigned long mask = ~(1L << id);
+  block->infly &= mask;
+}
+
+void req_callback(struct uk_blkreq *req, void *cookie)
+{
+  token_t *token = (token_t*)cookie;
+  block_t *block = token->block;
+
+  token_id_t tok_id = token_id(block, token);
+
+  signal_block_request_ready(block->id, tok_id);
+}
+
+void queue_callback(struct uk_blkdev *dev, uint16_t queue_id, void *argp)
+{
+  (void)argp;
+
+  int rc = uk_blkdev_queue_finish_reqs(dev, 0);
+  if (rc) {
+    uk_pr_err("Error finishing request: %d\n", rc);
+  }
+}
+
+static void init_token(token_t *token, bool write, unsigned long sstart,
+    unsigned long size, char *buf)
+{
+  struct uk_blkreq *req = &token->req;
+
+  uk_blkreq_init(req, write ? UK_BLKREQ_WRITE : UK_BLKREQ_READ, sstart, size,
+      buf, req_callback, token);
+
+  assert(!uk_blkreq_is_done(&token->req));
+}
+
+static long block_io(block_t *block, int write, unsigned long sstart,
```

Comment:
```suggestion
token_id_t block_io(block_t *block, int write, unsigned long sstart,
```
and setting `token_id_t` to signed `int` (or define a `stoken_id_t`?); and adjust the callers.